// gulp-apply-template/index.js
var engines = require('consolidate');
var es = require('event-stream');
var gutil = require('gulp-util');

const NAME = require('./package.json').name;


module.exports = plugin;
module.exports.cons = engines;


/**
 * Handler methods for Buffer and Stream contents.
 *
 * read(contents, cb):
 *   Given Stream/Buffer `contents`, return it as string
 *   via callback.
 *
 * convert(contents, cb):
 *   Given string `contents`, return it as the appropriate
 *   Stream/Buffer type via callback.
 */

var bufferHandler = {
  read: function (contents, cb) {
    cb(null, contents.toString());
  },
  convert: function (contents, cb) {
    cb(null, new Buffer(contents));
  }
};

var streamHandler = {
  read: function (contents, cb) {
    contents.pipe(es.wait(cb));
  },
  convert: function (contents, cb) {
    // Not sure which is more optimal: convert to a stream
    // of one big chunk, or split into chunks by newlines
    cb(null, es.readArray([contents]));
  }
};


/**
 * Transform Buffer or Stream file contents.
 *
 * file: the Vinyl object
 * transform: function(contentString, callback)
 * callback: function(err, file)
 */

function transform (file, transformer, callback) {
  var handler;

  if (file.isBuffer()) {
    handler = bufferHandler;
  }

  if (file.isStream()) {
    handler = streamHandler;
  }

  // Read file buffer/stream contents
  handler.read(file.contents, function (err, contents) {
    // Pass `contents` string to transformer function
    transformer(contents, function (err, transformed) {
      if (err) {
        callback(err);
      } else {
        // Convert back `transformed` string to buffer/stream
        handler.convert(transformed, function (err, converted) {
          file.contents = converted;
          callback(null, file);
        });
      }
    });
  });
}


/**
 * Create the plugin's through stream.
 *
 * options:
 *   engine: The template engine (required)
 *   template: The template file path (required)
 *   context: The default template context
 *   props: Array of names of Vinyl object properties to assign to context,
 *          defaults to [path, contents, data]
 */

function plugin (options) {
  options = options || {};

  var
    engine = options.engine,
    template = options.template,
    context = options.context,
    props = options.props || ['path', 'contents', 'data'];

  if (!engine) {
    throw new gutil.PluginError(NAME, NAME + ': No template engine supplied');
  }

  if (!template) {
    throw new gutil.PluginError(NAME, NAME + ': No template file supplied');
  }

  return es.map(function (file, callback) {
    var
      userEngine = engine,
      userTemplate = template,
      userContext = context || {},
      currentContext = {},
      contentsInTemplate = false;

    // Wrap `callback` to cast any errors as a PluginError
    function emit (err, file) {
      if (err) {
        err.message = NAME + ': ' + err.message;
        err = new gutil.PluginError(NAME, err, {showStack: true});
        callback(err);
      } else {
        callback(null, file);
      }
    }

    // Dynamically generate user context
    if (typeof userContext === 'function') {
      userContext = userContext(file);
    }

    // Copy specified file properties to the template context
    if (Array.isArray(props)) {
      props.forEach(function (key) {
        // If is a function, bind to `file`, so that we can
        // use methods like `file#inspect()` in the template
        // (not sure what the use case for is this though)
        if (typeof file[key] === 'function') {
          currentContext[key] = file[key].bind(file);
        } else {
          currentContext[key] = file[key];
        }
      });
    }

    // Copy over user context
    if (userContext) {
      Object.keys(userContext).forEach(function (key) {
        currentContext[key] = userContext[key];
      });
    }

    // Finally, expose the raw file object in the context,
    // this property cannot be overridden by the user.
    currentContext.file = file;

    // Flag indicating that the `contents` prop in the
    // context is the actual file contents, not something
    // that was overidden through the `options.context`
    contentsInTemplate = (
      currentContext.contents &&
      currentContext.contents === file.contents
    );

    // Dynamically determine what template engine to use
    if (typeof userEngine === 'function') {
      userEngine = userEngine(currentContext, file);
    }

    // Dynamically determine what template path to use
    if (typeof userTemplate === 'function') {
      userTemplate = userTemplate(currentContext, file);
    }

    if (file.isBuffer() || file.isStream()) {
      transform(file, function (contents, cb) {
        if (contentsInTemplate) {
          currentContext.contents = contents;
        }
        // console.log(currentContext);
        engines[userEngine](userTemplate, currentContext, cb);
      }, emit);
    } else {
      callback(null, file);
    }
  });
}
